File: gtci/09-hashmaps/python3/5-ransom-note.py
from collections import defaultdict


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        ransom_note_freq_map = {}
        for char in ransomNote:
            if char not in ransom_note_freq_map:
                ransom_note_freq_map[char] = 0
            ransom_note_freq_map[char] += 1

        magazine_freq_map = {}
        for char in magazine:
            if char not in magazine_freq_map:
                magazine_freq_map[char] = 0
            magazine_freq_map[char] += 1

        for char, count in ransom_note_freq_map.items():
            if char not in magazine_freq_map:
                return False

            if count > magazine_freq_map[char]:
                return False

        return True


if __name__ == "__main__":
    sol = Solution()

    assert sol.canConstruct("hello", "hellworld") == True
    assert sol.canConstruct("notes", "stoned") == True
    assert sol.canConstruct("apple", "pale") == False

    print("All test cases passed.")

File: gtci/09-hashmaps/cpp/5-ransom-note.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    bool canConstruct(string ransomNote, string magazine) {
        std::unordered_map<char, int> ransomNoteFreqMap;
        for (char ch : ransomNote) ransomNoteFreqMap[ch] += 1;

        std::unordered_map<char, int> magazineFreqMap;
        for (char ch : magazine) magazineFreqMap[ch] += 1;

        for (const auto& [ch, count] : ransomNoteFreqMap) {
            if (!magazineFreqMap.contains(ch)) return false;

            if (count > magazineFreqMap[ch]) return false;
        }

        return true;
    }
};

int main() {
    Solution sol;

    assert(sol.canConstruct("hello", "hellworld") == true);
    assert(sol.canConstruct("notes", "stoned") == true);
    assert(sol.canConstruct("apple", "pale") == false);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}
